File: dynamic_programming/knapsack_tab.js
function knapsack(profits, weights, capacity) {
    const dp = Array(profits.length).fill(0).map(() => Array(capacity + 1).fill(0))

    for (let c = 0; c < capacity + 1; c++) {
        if (c - weights[0] >= 0) {
            dp[0][c] = profits[0]
        }
    }

    for (let id = 1; id < profits.length; id++) {
        for (let c = 0; c < capacity + 1; c++) {
            const include = c - weights[id] >= 0
                ? profits[id] + dp[id - 1][c - weights[id]]
                : 0
            const exclude = dp[id - 1][c]
            dp[id][c] = Math.max(include, exclude)
        }
    }

    return dp[profits.length - 1][capacity]
}

function main() {
    console.log(knapsack([1, 6, 10, 16], [1, 2, 3, 5], 5))
    console.log(knapsack([1, 6, 10, 16], [1, 2, 3, 5], 6))
    console.log(knapsack([1, 6, 10, 16], [1, 2, 3, 5], 7))
}


main()

File: dynamic_programming/knapsack_tab.py
def knapsack(profits, weights, capacity):
    # Table with id as rows and capacity as columns
    dp = [[0 for _ in range(capacity+1)] for _id in range(len(weights))]
    
    # Setup first row of table, base condition where id is 0
    for c in range(capacity + 1):
        if c - weights[0] >= 0:
            # if capacity still exists, we can add the profit
            dp[0][c] = profits[0]

    for id in range(1, len(weights)):
        for c in range(1, capacity + 1):
            include = 0

            if c - weights[id] >= 0:
                # here we add the current profit, plus,
                # the max profit possible from the previous iterations at the remaining capacity
                include = profits[id] + dp[id - 1][c - weights[id]]
            
            exclude = dp[id - 1][c]

            dp[id][c] = max(include, exclude)

    return dp[len(weights) - 1][capacity]

def main():
  print(knapsack([1, 6, 10, 16], [1, 2, 3, 5], 5))
  print(knapsack([1, 6, 10, 16], [1, 2, 3, 5], 6))
  print(knapsack([1, 6, 10, 16], [1, 2, 3, 5], 7))


main()